include/deal.II/lac/read_write_vector.h

```diff
@@ -0,0 +1,715 @@
+     * Prints the vector to the output stream @p out.
+     */
+    void print (std::ostream       &out,
+                const unsigned int  precision  = 3,
+                const bool          scientific = true) const;
+
+    /**
+     * Returns the memory consumption of this class in bytes.
+     */
+    std::size_t memory_consumption () const;
+    //@}
+
+  protected:
+
+    /**
+     * Return the local position of @p global_index.
+     */
+    unsigned int
+    global_to_local (const types::global_dof_index global_index) const
+    {
+      // the following will throw an exception if the global_index is not
+      // in the remaining_elements
+      return static_cast<unsigned int>(stored_elements.index_within_set(global_index));
+    }
+
+    /**
+     * A helper function that is used to resize the val array.
+     */
+    void resize_val (const size_type new_allocated_size);
+
+    /**
+     * Indices of the elements stored.
+     */
+    IndexSet stored_elements;
+
+    /**
+     * Pointer to the array of local elements of this vector.
+     */
+    // TODO: Could we use VectorizedArray here for storage?
+    Number *val;
+  };
+
+  /*@}*/
+
+
+  /*----------------------- Inline functions ----------------------------------*/
+
+#ifndef DOXYGEN
+
+  template <typename Number>
+  inline
+  ReadWriteVector<Number>::ReadWriteVector ()
+    :
+    val (nullptr)
+  {}
+
+
+
+  template <typename Number>
+  inline
+  ReadWriteVector<Number>::ReadWriteVector (const ReadWriteVector<Number> &v)
+    :
+    Subscriptor(),
+    val (nullptr)
+  {
+    this->operator=(v);
+  }
+
+
+
+  template <typename Number>
+  inline
+  ReadWriteVector<Number>::ReadWriteVector (const size_type size)
+    :
+    val (nullptr)
+  {
+    reinit (size, false);
+  }
+
+
+
+  template <typename Number>
+  inline
+  ReadWriteVector<Number>::ReadWriteVector (const IndexSet &locally_stored_indices)
+    :
+    val (nullptr)
+  {
+    reinit (locally_stored_indices);
+  }
+
+
+
+  template <typename Number>
+  inline
+  ReadWriteVector<Number>::~ReadWriteVector ()
+  {
+    resize_val(0);
+  }
+
+
+
+  template <typename Number>
+  inline
+  ReadWriteVector<Number> &
+  ReadWriteVector<Number>::operator= (const ReadWriteVector<Number> &in_vector)
+  {
+    resize_val(in_vector.size());
+    stored_elements = in_vector.get_stored_elements();
+    std::copy(in_vector.begin(),in_vector.end(),begin());
+
+    return *this;
+  }
+
+
+
+  template <typename Number>
+  template <typename Number2>
+  inline
+  ReadWriteVector<Number> &
+  ReadWriteVector<Number>::operator= (const ReadWriteVector<Number2> &in_vector)
+  {
+    resize_val(in_vector.size());
+    stored_elements = in_vector.get_stored_elements();
+    std::copy(in_vector.begin(),in_vector.end(),begin());
+
+    return *this;
+  }
+
+
+
+  template <typename Number>
+  inline
+  typename ReadWriteVector<Number>::size_type
+  ReadWriteVector<Number>::size () const
+  {
+    return stored_elements.n_elements();
```

Comment:
No, `IndexSet::size()` is the correct answer here. We consider a `ReadWriteVector` as a vector of a certain size of which only a subset of elements are stored. The size of the vector is still the global size. So I think it's clear what this function should return. There could be another function, say `n_stored_elements()` (matching the function `get_stored_elements()` just below) that returns how many elements of the vector this class actually stores.